### ISSUE TITLE
Replace original link when re-adding via --nocopy

### ISSUE THREAD

**Issue by NatoBoram** [ipfs/go-ipfs#7465]
Context : https://github.com/ipfs/go-ipfs/issues/7373#issuecomment-639972271

# Problem

If a file previously added with `--nocopy` is modified then re-added, there's an error.

> Error: data in file did not match. mirrors/ubuntu/dists/focal-updates/InRelease offset 0

This can be especially annoying when trying to mirror a package manager to IPFS, because often they'll have an index file (named `InRelease` for Ubuntu) that will change over time.

# Workarounds

To get around this problem, filesystem snapshots can be used. This way, one can implement the following workflow : 

1. Host mirror files at `/mnt/storage/mirrors`
2. Create snapshots after downloading updates at `/mnt/storage/snapshots`
3. Add the snapshot to IPFS

Rinse and repeat. The problem with that is that the `/mnt/storage/snapshots` folder will get an ever-growing amount of snapshots to keep without any possibility to cleanup without braking the whole mirror. That can be a bit annoying, so we need a mechanism for IPFS to be a bit more machine-friendly when using `--nocopy`.

# Solutions

* If a file added using `--nocopy` matches another one using `--nocopy`, the old link should be replaced with the new link.

This will allow the removal of obsolete filesystem snapshots without breaking the mirror. This *could* be implemented with a `--replace` flag, but ideally, it should be the default behaviour.

* If a file added using `--nocopy` has the same path as another one but a different hash, everything using `--nocopy` that's associated with the old hash should be removed from the datastore.

In this situation, *the old hash is broken anyway*. It can be safely dropped and replaced. To facilitate the transition between a broken state and a non-broken state, `ipfs add --nocopy` should not break on inconsistent data. Instead, it should continue business as usual but remove the old hash and everything depending on it.

This could be implemented using a `--cascade` flag.

* If a file added using `--nocopy` has the same path as another one but a different hash, attempt to recover the broken file via IPFS and store it *without* using `--nocopy`.

It might occur that the modification to the original file was accidental or that we want to preserve the old file to keep the old mirror going. The file could even be recoverable from the IPFS network if someone else has it. In such case, it would make sense to attempt to promote broken `--nocopy` files.

This could be implemented using a `--recover` flag.

**Comment by Stebalien** [comment]
> If a file previously added with --nocopy is modified then re-added, there's an error.
>>    Error: data in file did not match. mirrors/ubuntu/dists/focal-updates/InRelease offset 0

Not quite. If a file is added then _changed_, there is an error. This isn't directly related to re-adding the file.

---

This solution isn't really an option, I think I misunderstood your comment in the other thread. Files added via the filestore can't be changed unless the blocks revering to these files have been removed.

One option would be to treat corrupt blocks in the filestore as simply "missing". However, this could have a significant performance impact on checking if we _have_ blocks in the filestore as we'd have to check the hash (although we may be able to get away with not doing this).

The real tricky part here is dealing with block deduplication. If we:

1. Add some files via the filestore.
2. Pin some other content.
3. Remove the files from disk.

We may end up removing blocks used by the newly pinned content. To really make this work, we'd likely need to store the content _again_ in the blockstore in step 2, but doing that in practice is non-trivial.
